device/pcie/pci_device.cpp

```diff
@@ -493,3 +499,83 @@ void PCIDevice::write_tlb_reg(uint32_t byte_addr, std::uint64_t value_lower, std
     }
     tt_driver_atomics::mfence(); // Otherwise subsequent WC loads move earlier than the above UC store to the TLB register.
 }
+
+
+bool PCIDevice::is_hardware_hung() {
+    volatile const void *addr = reinterpret_cast<const char *>(bar0_uc) + (get_architecture_implementation()->get_arc_reset_scratch_offset() + 6 * 4) - bar0_uc_offset;
+    std::uint32_t scratch_data = *reinterpret_cast<const volatile std::uint32_t*>(addr);
+
+    return (scratch_data == 0xffffffffu);
+}
+
+bool PCIDevice::reset_by_sysfs() {
+
+    const char *virtual_env = getenv("VIRTUAL_ENV");
+    if (virtual_env == nullptr)
+        return false;
+
+    std::string reset_helper_path = virtual_env;
+    reset_helper_path += "/bin/reset-helper";
+
+    std::string busid = std::to_string(info.pci_bus);
+
+    suspend_before_device_reset();
+
+    char *argv[3];
+    argv[0] = const_cast<char*>(reset_helper_path.c_str());
+    argv[1] = const_cast<char*>(busid.c_str());
+    argv[2] = nullptr;
+
+    pid_t reset_helper_pid;
+    if (posix_spawn(&reset_helper_pid, reset_helper_path.c_str(), nullptr, nullptr, argv, environ) != 0)
+        return false;
+
+    siginfo_t reset_helper_status;
+    if (waitid(P_PID, reset_helper_pid, &reset_helper_status, WEXITED) != 0)
+        return false;
+
+    if (reset_helper_status.si_status != 0)
+        return false;
+
+    resume_after_device_reset();
+
+    return true;
+}
+
+bool PCIDevice::reset_by_ioctl() {
+    struct tenstorrent_reset_device reset_device;
+    memset(&reset_device, 0, sizeof(reset_device));
+
+    reset_device.in.output_size_bytes = sizeof(reset_device.out);
+    reset_device.in.flags = 0;
+
+    if (ioctl(device_fd, TENSTORRENT_IOCTL_RESET_DEVICE, &reset_device) == -1) {
+        return false;
+    }
+
+    return (reset_device.out.result == 0);
+}
+
+bool PCIDevice::auto_reset_board() {
```

Comment:
I don't think UMD should be in the business of auto resetting anything.  UMD is not in a position to restore any the hardware to a known good state, so at best this is a convenience for the developer (won't have to run tt-smi).  At worst, it will complicate debugging efforts if user hangs a chip.

For what it's worth, I've never seen this work the way one would expect it to from reading the code.

Reply:
Makes sense to me. I'll remove the auto reset feature, we can bring it back if the need arises. I'll leave the hang detection though.